File: engine/runtime-parser/src/main/scala/org/enso/compiler/core/ir/expression/warnings/Unreachable.scala
package org.enso.compiler.core.ir
package expression
package warnings

import com.oracle.truffle.api.source.Source

/** Warnings for unreachable code. */
sealed trait Unreachable extends Warning {
  val location: Option[IdentifiedLocation]
}

object Unreachable {

  /** A warning for unreachable branches in a case expression.
    *
    * @param location the location of the unreachable branches
    */
  sealed case class Branches(
    override val location: Option[IdentifiedLocation]
  ) extends Unreachable {
    val atLocation =
      if (location.isDefined) {
        s" at location ${location.get}"
      } else {
        ""
      }

    override def message(source: Source): String =
      s"Unreachable case branches$atLocation."

    override def diagnosticKeys(): Array[Any] = Array(atLocation)
  }
}

File: engine/runtime-parser/src/main/java/org/enso/compiler/core/ir/expression/warnings/Unreachable.java
package org.enso.compiler.core.ir.expression.warnings;

import org.enso.compiler.core.ir.IdentifiedLocation;
import org.enso.compiler.core.ir.Warning;
import scala.Function1;

/** Warnings for unreachable code. */
public sealed interface Unreachable extends Warning {
  /**
   * A warning for unreachable branches in a case expression.
   *
   * @param identifiedLocation
   */
  record Branches(IdentifiedLocation identifiedLocation) implements Unreachable {

    @Override
    public String message(Function1<IdentifiedLocation, String> source) {
      return "Unreachable case branches" + atLocation();
    }

    private String atLocation() {
      String atLocation;
      if (location().isDefined()) {
        atLocation = " at location " + location().get();
      } else {
        atLocation = "";
      }
      return atLocation;
    }

    @Override
    public Object[] diagnosticKeys() {
      return new Object[] {atLocation()};
    }
  }
}
